File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const stack = [];
    const map = {
        '}': '{',
        ']': '[',
        ')': '('
    }
    
    for (const char of s) {
        if (!(char in map)) {
            stack.push(char);
        } else {
            if (stack.length === 0 || stack.pop() !== map[char]) return false;
        }
    }
    
    if (stack.length > 0) return false;
    return true;
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        Map<Character, Character> map = new HashMap<>();

        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');

        for (char p: s.toCharArray()) {
            if (map.containsKey(p)) {
                if (!st.isEmpty() && st.peek().equals(map.get(p))) {
                    st.pop();
                } else {
                    return false;
                }
            } else {
                st.push(p);
            }
        }

        return st.isEmpty();
    }
}